### ISSUE TITLE
Possible improvements for training ML models

### ISSUE THREAD

**Issue by mathematiguy** [TeHikuMedia/corpora#54]
Following our discussion on https://github.com/TeHikuMedia/waha-tuhi/issues/75#issuecomment-406912192 - I can think of a few ideas that could be quite useful.

At the moment, the first step of the prepare process is to build up a table locally according to the schema below, which is basically a snapshot of the API which gets partitioned into training and test sets. Possibly we will be able to move some of those parts into the API where they only need to be calculated once.

### Transcript Data Schema
```
- recording_id : int <-- suffices for a primary key (enforce this)
- sentence_id : int
- speaker_id : int
- transcript : str
- approves : int?
- upvotes : int?
- downvotes : int?
- downloaded : datetime
- holdout_speaker : bool
- holdout_sentence : bool
- normalized_transcript: str
- wav_hash : str
- wav_ok : bool
```

With the exception of `holdout_speaker` and `holdout_sentence`, each of these values really only need to be calculated once for each transcript. With that said, deciding whether a given speaker or sentence is heldout is actually done by hashing the `speaker_id` or `sentence_id` respectively into a floating point from 0 to 1, and thresholding that as a means of random sampling.

Because of this, we could attach the seeds to the API and set the thresholds in the config file for training. So there really isn't a reason why the columns couldn't all be included in the API, technically speaking.

Happy to discuss on Tuesday by video conference.

**Comment by kmahelona** [comment]
Currently, the Recording API gives back,
```
{
    "person": 1,
    "sentence": {
        "id": 323,
        "text": "I konā rā ahau e whēke ana kia kake ahau i te puke.",
        "language": "mi",
        "quality_control": [...]
        "updated": "2017-09-27T11:30:01.823402+13:00",
        "source": null
    },
    "audio_file_url": "https://koreromaori.com/recording-file/551/",
    "quality_control": [...],
    "id": 551,
    "sentence_text": "I konā rā ahau e whēke ana kia kake ahau i te puke.",
    "user_agent": null,
    "created": "19-12-17 16:38 NZDT"
}
```
What we can do is add some aggregates to this such as,
```
    "quality_control_aggregates": {
        "approves": 1,
        "good": 3,
        "bad": 0,
        ...
    }
```

This would save you having to do the maths on the nested `quality_control` model.

Then the missing piece is to add a transcription field to this API, and I think that should be a Transcription object that has the same recording as a ForeignKey. I'll have to look at the Transcription model and see how this can work.

**Comment by edwardabraham** [comment]
That's great. Just clarifying. If the sentence is edited, does the `sentence` change, or the `sentence_text`.

The key requirement is to be able to  download all data where either the sentence or the quality control has changed. This would be achieved by adding  in an updated time to the quality control aggregate. (and making sure that these update times could be accessed via query parameters). 

The strategy then would be to find the most recently updated data in the data that we already had, and look for anything that was either missing, or had been updated more recently than the most recent update.

**Comment by kmahelona** [comment]
`sentence_text` is the field that changes as it's a field on the recording itself. `sentence` is a ForeignKey relation to the Sentence model. So when someone edits the `sentence_text` for a recording, the `updated` field on `recording` is automatically updated. 

We've had instances where after we got recordings on a `Sentence`, we corrected the sentence even after being approved. This doesn't change the `sentence_text` because the recording would have been on the old sentence, which is what's in `sentence_text`. I don't think changes on `Sentence` are worth considering for this.

**Comment by mathematiguy** [comment]
Following up on our discussion in the campfire, from the schema I described above, here are the fields that will need to be added:
- `last_modified` : datetime
- `holdout_speaker` : bool
- `holdout_sentence` : bool
- `wav_hash` : str
- `wav_ok` : bool

### Notes on holdout
`holdout_speaker` and `holdout_sentence` could be done in multiple ways. As explained above we hash the `speaker_id` to get a random number `speaker_seed` from 0 to 1. Then when we go to sample we get all of the speakers whose seed is below a low threshold, and leave them out of the training. 

The hashing makes sure that each speaker gets heldout appropriately, based only on their `speaker_id` which (hypothetically) never changes. We could use the code I have for id hashing to get seeds for each speaker, then they get a uniformly random seed as well as an id.

Alternatively, we could calculate a random number each time a new speaker is created. But then it becomes important to maintain the consistency of that random number. We don't want people bouncing in and out of the training set if we can help it, it's confusing.

## Wav checking

`wav_ok`: Because some wav files are either too short or improperly encoded for DeepSpeech to read them, we have to check the wav files before running the model. It would be best if we could do this as few times as possible, because it's an expensive operation (usually takes a few hours to run over all 200,000 recordings). 

I could provide some code to do this checking at the API end when the recording is sent - or possibly to queue a job to calculate this when the resources are available.

At the same time, I've also noticed (though it could be a problem with our code) that there is a small minority of wav files that don't successfully download, and either arrive blank or with some xml saying something like `download failed`. 

`wav_hash`: So on our end there might still be a need for some kind of checking, which is why an md5 hash of each wav file would be really handy. I haven't tried to query for the wav hash from S3 yet, but [this discussion](https://stackoverflow.com/questions/1775816/how-to-get-the-md5sum-of-a-file-on-amazons-s3) suggested to me that it might not be as simple as grabbing the ETag header.

If we can clarify whether we can grab the md5sum from the ETag header, that would be great.

**Comment by kmahelona** [comment]
- last modified
That’s fine. It’s “updated” in the model.

- hold out speaker/sentence
I don’t think we should put this in corpora - well we shouldn’t associate that generically with a recording/person. We could look at making a way for _you_ to save that sort of info in corpora so for example you might have a set of preferences that say what kinda data you want, but that’s making a lot more work than necessary atm. 

- wav_hash
S3 etag is the md5 hash in our case. I’ll add this to the mode though since it won’t “cost” much. 

- wav_ok
That’s like a quality control. If you’ve come across errors with wav files, then you should post a QC on the recoding saying it’s bad or even flagging it for deletion. That would be the preferred method. On that note we’ve added a number of additional indicators to the quality control so be sure to check that. 

- people IDs
These are always unique and won’t change. We actually have a UUID field for people as well so I could expose that for you to use. But I can see why you have the seed and go from 0-1 so I’m not sure giving you a UUID would add any extra benefit - unless you sorted by that and took a portion of the list.

**Comment by mathematiguy** [comment]
Thanks Keoni, really useful feedback.

### Response

- last modified
Awesome, I'll use 'updated' from now on.

- hold out speaker/sentence
I don't really mind keeping this out of corpora since it is so straightforwardly associated with the `speaker_id`/`sentence_id`, so that's easy.

- wav_hash
I don't mind grabbing the S3 etag at all, I just have to figure out how to do it 

- wav_ok 
I like the idea of posting QC to the API for bad recordings, esp if I can check the hash before I do it first. Let's go with that plan, I guess we just need a field added to indicate if the quality is poor. 
We could also consider a field that might explain why the file is bad as well.

- people IDs
Sorting the UIDs is a valid strategy for sampling so long as the UIDs are random. At the moment the code we have works, so it's ok not to change it.

### Summary
It looks like having a field I can post to for quality control is a good idea. I'll also figure out how to grab the etag from S3. The rest is ok as is, so far as I can tell.

**Comment by kmahelona** [comment]
- wav_hash
This will be added as `audio_file_md5`. It will take some time to ensure I do this correctly and that all ~200k recordings are updated. I'm testing this on our staging server atm.

- wav_ok
I'd suggest just sending a QC with `delete=True`. If the file doesn't exist, something happened along the way (we had bugs in the beginning), so just flag for deletion. We haven't implemented anything yet that deletes recordings, `delete=True` is just a flag until we decide when/how to delete recordings. So on that note make sure you don't include recordings with `delete=True` in your training set.

- people ids
Cool, I'll keep this as it is. One option is the `person` field could return a dictionary instead of an ID and I could put the UUID and other anonymous attributes in there. That would change the current API in a way that would affect you, and we currently don't have any API versioning.

### Posting a QualityControl
Check out https://koreromaori.com/api/qualitycontrol/ and https://koreromaori.com/docs/#api-qualitycontrol-create for how to post a QC object. `object_id` is the ID of the recording, and `content_type` will be the ID for the Recording model which is `8`. I should update that to take a string e.g. 'recording'...

**Comment by mathematiguy** [comment]
Great work Keoni, I've split the `download-audio` script into `download-json` and `download-audio` separately, and added the extra fields with placeholders on my end.

I'm curious if there's a way to request only the data where the `updated_after` field is greater than a pre-specified datetime. 

At the moment, if we have the `.wav` files downloaded already, we can check to see if we need to download new ones only if we haven't got something downloaded already. It would neat if we could do the same with the `.json`, since at the moment we can only download everything (or just 'approved' etc).

This sounds like something that could be handled by the postgres back end you mentioned earlier.

**Comment by kmahelona** [comment]
Hey @mathematiguy ,

The `?updated_after=DATETIME` will return everything with `updated >= DATETIME`. I guess for clarity we could use `updated__gte=` which is more like a Django query. Here's the "docs" https://dev.koreromaori.com/docs/#api-recordings-list (I think I just noticed a type). 

So something like,

```https://dev.koreromaori.com/api/recordings/?updated_after=2018-07-27T19:10:00%2B1200```

would give you a JSON with all the recording objects updated after (or equal to) that date. It's still paginated because the results could be huge. I'm just thinking that we could also do some generalised filters, e.g. ?is_approved or ?is_good (good > bad) if that could be useful.

**Note** I haven't pushed this to production yet because I'm still making sure I don't stuff anything up with large, "automated" database changes. The dev server has HEAPS of "missing files" because of how I've migrated things over for testing, so it's been useful in terms of ensuring I do things correctly. I'm hoping to get that done tonight so I'm not slowing you down ;).

**Comment by mathematiguy** [comment]
That's exactly what I had in mind! Looking forward to it.

> I'm hoping to get that done tonight so I'm not slowing you down ;).

I have lots of other stuff to work on in the meantime, no problem if you need more time :)

**Comment by kmahelona** [comment]
Oops, looks like it's still transcribing. It's done about 33k. There's about another 116k to go.

**Comment by kmahelona** [comment]
#### Transcription field
Yes, we're using DeepSpeech. These are still being transcribed. As mentioned above this will only happen for non-reviewed recordings.

#### Storing DS transcriptions
Yes, you can see them in the Django admin here https://koreromaori.com/admin/transcription/transcription/. I'll work on getting them in the API.

#### Rating
That's a good idea. You *could* use the "star" field. Since this is a Machine QC, if you give the star field anything from 1-5 and a note we'll know what this means. Currently the "Star" button on the UI for human reviewers is to indicate that something is good. I haven't made this into a number of stars rating, but that could happen easily.

**Comment by kmahelona** [comment]
Just fixed a typo (https://github.com/TeHikuMedia/corpora/commit/832361dc018326c31a8dc06a661277c57eccc1a7) so unfortunately we won't have logs for some 50k transcriptions. Good thing I checked! 😅

**Comment by kmahelona** [comment]
@mathematiguy I was just thinking, maybe we should do a `good=1` instead of `approved=True` when `transcription==sentence_text`. Because technically they read what is said but their pronunciation or inflection may not be perfect. This kinda makes me think, maybe we should train a proficient model and use that to automatically review all the beginner corpus. But we can think about that later, for now we'll stick with the current plan.

**Comment by kmahelona** [comment]
Everything's transcoded!
